File: apps/backend/internal/botengine/builtin.go
package botengine

import (
	"fmt"
	"math/rand"
	"strings"
)

// executeBuiltinHandler 执行内置事件
// Step 2 中会扩展完善，当前实现基础功能
func executeBuiltinHandler(config map[string]any, input string, variables map[string]string) (string, error) {
	builtinType := getStringField(config, "builtin_type")

	switch builtinType {
	case "random_number":
		return builtinRandomNumber(config)
	case "haiku":
		return builtinHaiku(config, input)
	case "echo":
		return builtinEcho(config, input)
	case "count":
		return builtinCount(config, variables)
	case "template":
		return builtinTemplate(config, input, variables)
	default:
		return "", fmt.Errorf("unknown builtin type: %s", builtinType)
	}
}

// builtinRandomNumber 生成随机数
func builtinRandomNumber(config map[string]any) (string, error) {
	minVal := 0
	maxVal := 100
	isInteger := true

	if v, ok := config["min"].(float64); ok {
		minVal = int(v)
	}
	if v, ok := config["max"].(float64); ok {
		maxVal = int(v)
	}
	if v, ok := config["integer"].(bool); ok {
		isInteger = v
	}

	if minVal > maxVal {
		minVal, maxVal = maxVal, minVal
	}

	if isInteger {
		return fmt.Sprintf("%d", rand.Intn(maxVal-minVal+1)+minVal), nil
	}
	return fmt.Sprintf("%f", rand.Float64()*float64(maxVal-minVal)+float64(minVal)), nil
}

// builtinHaiku 生成俳句（模板化）
func builtinHaiku(config map[string]any, input string) (string, error) {
	topic := getStringField(config, "topic")
	if topic == "" {
		topic = input
	}

	// 简单的俳句模板（5-7-5 结构）
	haikus := []string{
		fmt.Sprintf("%s的光芒\n照亮了前行的路\n脚步不停歇", topic),
		fmt.Sprintf("%s轻声吟唱\n风中传来回响\n万物皆有灵", topic),
		fmt.Sprintf("静观%s变\n一叶落而知秋至\n心如止水", topic),
		fmt.Sprintf("%s如流水\n昼夜不息向前\n奔向大海", topic),
	}

	return haikus[rand.Intn(len(haikus))], nil
}

// builtinEcho 回显输入
func builtinEcho(config map[string]any, input string) (string, error) {
	prefix := getStringField(config, "prefix")
	suffix := getStringField(config, "suffix")

	result := input
	if prefix != "" {
		result = prefix + result
	}
	if suffix != "" {
		result = result + suffix
	}

	return result, nil
}

// builtinCount 统计消息计数
func builtinCount(config map[string]any, variables map[string]string) (string, error) {
	counterKey := getStringField(config, "counter_key")
	if counterKey == "" {
		counterKey = "message_count"
	}

	count := 0
	if val, ok := variables[counterKey]; ok {
		_, _ = fmt.Sscanf(val, "%d", &count)
	}
	count++

	variables[counterKey] = fmt.Sprintf("%d", count)
	return fmt.Sprintf("%d", count), nil
}

// builtinTemplate 模板渲染
func builtinTemplate(config map[string]any, input string, variables map[string]string) (string, error) {
	template := getStringField(config, "template")
	if template == "" {
		return "", fmt.Errorf("template is empty")
	}

	result := template
	for key, value := range variables {
		result = strings.ReplaceAll(result, "{"+key+"}", value)
	}

	// 替换 args 变量（{args}, {args:N}）
	result = ReplaceArgsVars(result, input)

	return result, nil
}

File: packages/workflow-engine/src/nodes/builtin.ts
import { z } from 'zod';
import type { NodeDefinition } from '../types.js';
import { resolveTemplate, type ResolveContext } from '../resolver.js';

const builtinConfigSchema = z.object({
  builtin_type: z.enum(['random_number', 'haiku', 'echo', 'count', 'template']),
  min: z.number().optional(),
  max: z.number().optional(),
  integer: z.boolean().optional().default(true),
  topic: z.string().optional(),
  prefix: z.string().optional(),
  suffix: z.string().optional(),
  counter_key: z.string().optional(),
  template: z.string().optional(),
});

export const builtinNode: NodeDefinition<z.infer<typeof builtinConfigSchema>> = {
  type: 'builtin',
  label: '内置',
  category: 'processing',
  icon: '⚙',
  configSchema: builtinConfigSchema,
  async execute(input, config, ctx) {
    const cfg = config as z.infer<typeof builtinConfigSchema>;
    let output = '';

    switch (cfg.builtin_type) {
      case 'random_number':
        output = builtinRandomNumber(cfg);
        break;
      case 'haiku':
        output = builtinHaiku(cfg, input.rawInput);
        break;
      case 'echo':
        output = builtinEcho(cfg, input.rawInput);
        break;
      case 'count':
        output = builtinCount(cfg, ctx.variables);
        break;
      case 'template':
        output = builtinTemplate(cfg, ctx);
        break;
      default:
        throw new Error(`Unknown builtin type: ${cfg.builtin_type}`);
    }

    return {
      ports: {
        out_output: output,
        out_exec: 'true',
      },
    };
  },
};

function builtinRandomNumber(config: z.infer<typeof builtinConfigSchema>): string {
  const min = config.min ?? 0;
  const max = config.max ?? 100;
  const isInteger = config.integer ?? true;

  const lo = Math.min(min, max);
  const hi = Math.max(min, max);

  if (isInteger) {
    return String(Math.floor(Math.random() * (hi - lo + 1)) + lo);
  }
  return String(Math.random() * (hi - lo) + lo);
}

function builtinHaiku(config: z.infer<typeof builtinConfigSchema>, input: string): string {
  const topic = config.topic || input;

  const haikus = [
    `${topic}的光芒\n照亮了前行的路\n脚步不停歇`,
    `${topic}轻声吟唱\n风中传来回响\n万物皆有灵`,
    `静观${topic}变\n一叶落而知秋至\n心如止水`,
    `${topic}如流水\n昼夜不息向前\n奔向大海`,
  ];

  return haikus[Math.floor(Math.random() * haikus.length)]!;
}

function builtinEcho(config: z.infer<typeof builtinConfigSchema>, input: string): string {
  let result = input;
  if (config.prefix) result = config.prefix + result;
  if (config.suffix) result = result + config.suffix;
  return result;
}

function builtinCount(
  config: z.infer<typeof builtinConfigSchema>,
  variables: Record<string, string>,
): string {
  const key = config.counter_key || 'message_count';
  const current = parseInt(variables[key] || '0', 10) || 0;
  const next = current + 1;
  variables[key] = String(next);
  return String(next);
}

function builtinTemplate(
  config: z.infer<typeof builtinConfigSchema>,
  ctx: ResolveContext,
): string {
  const template = config.template;
  if (!template) throw new Error('template is empty');

  // 统一变量替换（${path} + 所有遗留格式）
  return resolveTemplate(template, ctx);
}
